fix starting issues

bernieke opened the pull request:
Starting with the recommended privilege drops fails because of:
* root not being able to chown once /data is no longer owned by root
* git clone needs to be executed as radicale once /data is no longer root owned
* /etc/gitconfig is read-only
* git clone fails if the repo has already been cloned (existence check of the .git folder needs to be done by radicale as well since root can't even read /data, which is why I put the radicale exec stuff in a separate shell script instead of using a bunch of su-exec for every command)

I couldn't really figure out the logic of the thing (when would the entrypoint script not be executed by root for instance). So if I've done it all wrong, please let me know the logic, and I'll fix it the "proper" way.

tomsquest commented:
Hi @bernieke ,

Thanks for you PR.

I am thinking about that since the time you opened it.
There are a lot of cases to handle (is the git repo existing, owned by radicale, writable by radicale, cloned with ssh, is the git config working), and as you saw the git config is global instead of being local (easy fix, just get rid of the `--global` flag).

And for all of these, I am thinking that cloning the git repo and setuping it with the user/email stuff has to be done only once.
Thus, it can be done by the user **outside* the image, ie. the image has nothing to do with git repo/cloning stuff.

So I will take a bit more time, but maybe the solution will be to remove the support of "automatic git repo cloning" from the image.
Less maintenance, less complex image, less bug (like those one).

bernieke commented:
> > The fix for chown is still needed though, due to the dropped permissions root can't chown a folder once it's no longer owned by root (the second time the container is started).
> 
> Hum, I think `root` can always chown.
> 
> Can you provide instructions to make it fail ?

```
$ docker run --name radicale --read-only --init --security-opt="no-new-privileges:true" --cap-drop ALL --cap-add CHOWN --cap-add SETUID --cap-add SETGID --cap-add KILL --pids-limit 50 --memory 256M -v temp:/data  tomsquest/docker-radicale
chown: /data: Permission denied
```
 
> > Files owned within the volume are the users responsibility. If they're created by radicale they'll have the proper ownership, if the user created them themselves (git clone for instance) they should just chown them as well.
> 
> Actually, no manual action is needed: the `/data` directory will have its permission set to radicale on start.

True, I missed that bit in the Dockerfile.

So the chown (without UID/GID change) can just be left out altogether from the entrypoint then?

> > Also the gitconfig location needs to be fixed in some way since nothing except /data is writable.
> 
> With the removal of `--system` in `git config --system ...`, the gitconfig will reside in the $GIT_REPO/config.
> So there is no need to symlink or handle the global gitconfig file.

Also true, I was thinking the change would go into ~/.gitconfig instead of in <repo>/.git/config (I've only recently moved from mercurial to git...)

So then this bit can be removed as well, since if the user is responsible for checking out the repository, they can configure it as well? I've modified the README.md to reflect this.

> > Come to think of it, I think the UID/GID stuff breaks as well with the readonly permissions, since you can't write to /etc/passwd or /etc/group either... But I have absolutely no idea on how this could even be fixed, functionality like this seems completely incompatible with the readonly concept?
> 
> Yes, exactly : custom UID/GID require changes on files that are readonly.
> It is documented here: https://github.com/tomsquest/docker-radicale#option-2-custom-usergroup-at-run-time
> 
> > The only solution there I can think of would be to simply throw an error when a UID/GID change is required but the system is readonly.
> 
> That would be good idea for a PR with just that.

Implemented. I've also added instructions to the readme on how to solve the issue by temporarily dropping the restrictions.

Timost commented:
Hi everyone,
Thank you this docker image and for submitting these fixes.
A small comment to say, that I have the same `chown: /data: Permission denied` issue using the recommended settings (dropping capabilites etc.).
This happens *the second time* the container is started with an existing `data` volume which has had it's permission changed during the first run.
My current workaround is to change the owner and group back to the `root` user before starting the container again.

tomsquest commented:
Thanks for your comments.

I will try to reproduce the `permission denied` problem here.

I got rid of the git cloning step in https://github.com/tomsquest/docker-radicale/commit/3e4e286f21e9fc4d4a0e380522d8db98f7fa9c74
Instructions have been added to the README.

tomsquest commented:
@bernieke I saw a lot of modification (markdown spacing, instructions, detect readonly fs).
I will split these in separate commits ASAP.

tomsquest commented:
@bernieke added your check of read-only fs in ac3d88af7578cc5a3552b0398d93a9a43ad77b49

tomsquest commented:
Tried to reproduce the problem of permission, with the production config, using `latest` (master, same as 3.0.6.1).
Everything seems fine.

Here are the steps:

```bash
# Clean
sudo rm -rf /radicale

# Create the directories where the data and the config will be stored
sudo mkdir -p /radicale/{config,data}

# copy the default config from this repo
sudo cp config /radicale/config/
```

So here, we have everything.


Here is the initial permissions.
See that everything is own by `root` because we created the directories and the `config` file with `root`.

```
$ tree -pug /radicale
/radicale
├── [drwxr-xr-x root     root    ]  config
│   └── [-rw-r--r-- root     root    ]  config
└── [drwxr-xr-x root     root    ]  data
```

Let's start the container:

```bash
$ docker run --name radicale \
    -p 127.0.0.1:5232:5232 \
    --read-only \
    --init \
    --security-opt="no-new-privileges:true" \
    --cap-drop ALL \
    --cap-add CHOWN \
    --cap-add SETUID \
    --cap-add SETGID \
    --cap-add KILL \
    --pids-limit 50 \
    --memory 256M \
    --health-cmd="curl --fail http://localhost:5232 || exit 1" \
    --health-interval=30s \
    --health-retries=3 \
    -v /radicale/data:/data \        
    -v /radicale/config:/config:ro \
    tomsquest/docker-radicale
Unable to find image 'tomsquest/docker-radicale:latest' locally
latest: Pulling from tomsquest/docker-radicale
8464c5956bbe: Pull complete 
0590fa81014b: Pull complete 
60340def18a2: Pull complete 
b15f6116111f: Pull complete 
83385d540226: Pull complete 
a033b8cc36ee: Pull complete 
d90eb3a01ca1: Pull complete 
f62090b74cdc: Pull complete 
5ae955211910: Pull complete 
5859b5221080: Pull complete 
6e11fcad0319: Pull complete 
Digest: sha256:9e7fcb22bcda8d5af7d2dea01d4f361c982572d6f151de299723f285f287285a
Status: Downloaded newer image for tomsquest/docker-radicale:latest
```

The container is started.
The UI is accessible at http://localhost:5232/
I created a calendar (and used the user `tom` without the password).

Then I stopped the container (ctrl+c).
And started it again.
The container started successfully. It seems that this is where there is some permission issues that I can't reproduce.

Here are the permission of the files at this point :
```
$ tree -pug /radicale
/radicale
├── [drwxr-xr-x root     root    ]  config
│   └── [-rw-r--r-- root     root    ]  config
└── [drwxr-xr-x 2999     2999    ]  data
    └── [drwxr-xr-x 2999     2999    ]  collections
        └── [drwxr-xr-x 2999     2999    ]  collection-root
            └── [drwxr-xr-x 2999     2999    ]  test
                └── [drwxr-xr-x 2999     2999    ]  7b705b9a-7ae8-59e3-0711-86700fd05b82
```

See the permission 2999 meaning the container chmod'ed the filtes on start.

Timost commented:
Thank you for looking into it.. I've just tested with `3.0.6.1` I still get the same permission error. I'll try to make detailed instructions to reproduce.

bernieke commented:
Well, I found the issue: the chown works with bind mounts (`-v /host-path:/data`) but not with volumes (`-v volume-name:/data`).

May I suggest just having the chown fail silently? I don't really see any other way to both keep supporting the chown and volumes.

I've adjusted my PR accordingly.

tomsquest commented:
You nailed it : `chroot` of volumes (not bind-mount) is not working.
I tried to search for that without luck so far.

> May I suggest just having the chown fail silently? I don't really see any other way to both keep supporting the chown and volumes.

Yeah. I still hope to find a solution. Redis image uses `chroot` (and some others also).
I also know Elasticsearch docker image that provide a `TAKE_FILE_OWNERSHIP=true` env var.
Maybe this could be the default and volume users would have to make it `false`.
Failing silently, or with a warning message... hum I have to think about that :-/

Oh btw, @bernieke , I saw that your PR changes the permission checks in the python tests. Why so ?

bernieke commented:
> Oh btw, @bernieke , I saw that your PR changes the permission checks in the python tests. Why so ?

It fixes these otherwise failing tests:
```
> pytest -v                                                  (docker-radicale) 
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-6.1.2, py-1.9.0, pluggy-0.13.1 -- /home/bernieke/.virtualenvs/docker-radicale/bin/python
cachedir: .pytest_cache
rootdir: /home/bernieke/git/bernieke/docker-radicale
plugins: testinfra-5.3.1
collected 12 items                                                             

test_image_custom_build.py::test_process PASSED                          [  8%]
test_image_custom_build.py::test_port PASSED                             [ 16%]
test_image_custom_build.py::test_version PASSED                          [ 25%]
test_image_custom_build.py::test_user PASSED                             [ 33%]
test_image_custom_build.py::test_config_readonly FAILED                  [ 41%]
test_image_custom_build.py::test_data_folder_writable PASSED             [ 50%]
test_image_prod.py::test_process PASSED                                  [ 58%]
test_image_prod.py::test_port PASSED                                     [ 66%]
test_image_prod.py::test_version PASSED                                  [ 75%]
test_image_prod.py::test_user PASSED                                     [ 83%]
test_image_prod.py::test_config_readonly FAILED                          [ 91%]
test_image_prod.py::test_data_folder_writable PASSED                     [100%]

=================================== FAILURES ===================================
_____________________________ test_config_readonly _____________________________

host = <testinfra.host.Host docker://2ed4ff45c298248fe6e082d76d2fead2b2026f28f5c3df71cf4a3040542f0b75>

    def test_config_readonly(host):
        config_file = '/config/config'
        assert host.file(config_file).user == 'root'
        assert host.file(config_file).group == 'root'
>       assert host.file(config_file).mode == 0o664
E       assert 420 == 436
E         +420
E         -436

test_image_custom_build.py:44: AssertionError
_____________________________ test_config_readonly _____________________________

host = <testinfra.host.Host docker://3033e09538aa94b18fed576191de00389598bb9e92caf7f7bb80a62d8518bf7e>

    def test_config_readonly(host):
        config_file = '/config/config'
        assert host.file(config_file).user == 'root'
        assert host.file(config_file).group == 'root'
>       assert host.file(config_file).mode == 0o664
E       assert 420 == 436
E         +420
E         -436

test_image_prod.py:54: AssertionError
=========================== short test summary info ============================
FAILED test_image_custom_build.py::test_config_readonly - assert 420 == 436
FAILED test_image_prod.py::test_config_readonly - assert 420 == 436
========================= 2 failed, 10 passed in 5.37s =========================
```

bernieke commented:
> Yeah. I still hope to find a solution. Redis image uses `chroot` (and some others also).
> I also know Elasticsearch docker image that provide a `TAKE_FILE_OWNERSHIP=true` env var.
> Maybe this could be the default and volume users would have to make it `false`.
> Failing silently, or with a warning message... hum I have to think about that :-/

If you want to go with the TAKE_FILE_OWNERSHIP=true env var, let me know and I'll gladly implement it like this.

tomsquest commented:
Hi @bernieke 

I pushed a new PR with a new `TAKE_FILE_OWNERSHIP` env var.
This will allow to skip the `chown` when set to `false`.

Can you try the PR #91 ?

bernieke commented:
All starting issues have been handled.

Timost commented:
Thank you for fixing this ! :pray: :muscle:  `3.0.6.2` fixes the issue for me too.
